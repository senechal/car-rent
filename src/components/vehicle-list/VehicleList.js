
import { SORT_OPTION } from './constants'
import { Tag } from '@reusable';
import Vehicle from '@components/vehicle';
import { removeFilter } from '@components/search-filter/actions';
import { useDispatch, useSelector } from 'react-redux';
import { changeSortBy } from './actions';
import { useCallback } from 'react';

const { Group } = Tag;

const mergeFilters = (filters) => {
  return filters.reduce((acc, filter) => {
    const { field } = filter;
    if(acc[field]){
      acc[field] = [...acc[field], filter];
      return acc
    }
    acc[field] = [filter];
    return acc;
  }, {});
};

const sortVehicles = (sortBy, order) => (vehA, vehB) => {
  const isAsc = order === 'asc';

  const {
    [sortBy]: valA,
  } = vehA;
  const {
    [sortBy]: valB,
  } = vehB;

  if (parseInt(valA, 10) < parseInt(valB, 10)) {
    return isAsc ? -1 : 1;
  }
  if (parseInt(valA, 10) > parseInt(valB, 10)) {
    return isAsc ? 1 : -1;
  }

  return 0;
};

const VehicleList = () => {

  const {
    filters,
    vehicles,
    sortBy: {
      key: sortBy,
      order,
    },
  } = useSelector(state => state);

  const dispatch = useDispatch();

  const mergedFilters = mergeFilters(filters)

  const filteredVehicles = Object.keys(mergedFilters).reduce((acc, field) => {
    const fieldFielters = mergedFilters[field];
    return acc.filter(({[field]: key}) => {
      return fieldFielters.some(toFilter => {
        return key === toFilter.value;
      });
    });
  }, vehicles);

  const sortedVehicles = filteredVehicles.sort(sortVehicles(sortBy, order));

  const handleSortByChange = useCallback((event) => {
    const { value } = event.target;
    const [eventSortBy, eventOrder] = value.split(':');
    dispatch(changeSortBy(eventSortBy, eventOrder));
  }, [dispatch]);

  const handleRemoveFilter = useCallback((tagValue) => {
    const [field, value] = tagValue.split(':');
    dispatch(removeFilter(field, value));
  }, [dispatch]);

  return (
    <>
      <section className="rac-vehicle-list-header">
        <Group
          onClose={handleRemoveFilter}
          className="rac-vehicle-list-applied-filters"
        >
          Filters applied: {filters.map(({value, field, label}) => (
            <Tag value={`${field}:${value}`} key={label}>{label}</Tag>
          ))}
        </Group>
        <section className="rac-vehicle-list-sort-by">
          <label> Sort By: </label>
          <select onChange={handleSortByChange} value={`${sortBy}:${order}`}>
            {SORT_OPTION.map(({label, sortBy:optSortBy, order:optOrder }) => {
              return (
                <option value={`${optSortBy}:${optOrder}`} key={label}>{label}</option>
              )
            })}
          </select>
        </section>
      </section>
      {
        sortedVehicles.map(vehicle => {
          const { id } = vehicle;
          return <Vehicle key={id} {...vehicle} />
        })
      }
    </>
  );
};

export default VehicleList;
