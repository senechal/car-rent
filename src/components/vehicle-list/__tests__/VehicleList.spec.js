import { useSelector } from 'react-redux';
import { render, fireEvent } from '@testing-library/react';
import VehicleList from '../VehicleList';
import {
  CHANGE_SORT_BY,
  REMOVE_FILTER,
} from '@utils/actionTypes';

const mockVehicles = [
  {
    baggageQuantity: '2',
    currencyCode: 'CAD',
    doors: '2',
    fuelType: 'Petrol',
    hasAirConditioning: true,
    id: 'HERTZ_1',
    imgUrl: 'https://ctimg-fleet.cartrawler.com/kia/rio/primary.png?auto=format&w=600',
    makeModel: 'Kia Rio or similar',
    price: '731.47',
    seats: '5',
    status: 'Available',
    transmissionType: 'Automatic',
    vendor: 'HERTZ',
  },
  {
    baggageQuantity: '4',
    currencyCode: 'CAD',
    doors: '4',
    fuelType: 'Petrol',
    hasAirConditioning: true,
    id: 'AVIS_0',
    imgUrl: 'https://ctimg-fleet.cartrawler.com/chevrolet/impala/primary.png?auto=format&w=600',
    makeModel: 'Chevrolet Impala or similar',
    price: '836.25',
    seats: '5',
    status: 'Available',
    transmissionType: 'Automatic',
    vendor: 'AVIS',
  },
  {
    baggageQuantity: '4',
    currencyCode: 'CAD',
    doors: '4',
    fuelType: 'Petrol',
    hasAirConditioning: true,
    id: 'ALAMO_1',
    imgUrl: 'https://ctimg-fleet.cartrawler.com/chrysler/300/primary.png?auto=format&w=600',
    makeModel: 'Chrysler 300 or similar',
    price: '851.03',
    seats: '5',
    status: 'Available',
    transmissionType: 'Automatic',
    vendor: 'ALAMO',
  },
];

const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

jest.mock('react-router-dom', () => ({
  Link: ({children, to}) => <a href={to}>{children}</a>,
}));

describe('Testing VehicleList', () => {
  beforeEach(() => {
    useSelector.mockReset();
    mockDispatch.mockReset();
  });
  it('should render correctly', () => {
    useSelector.mockReturnValue({
      filters: [],
      vehicles: mockVehicles,
      sortBy: {
        key: 'price',
        order: 'asc',
      },
    });
    const { container } = render(<VehicleList />);
    expect(container).toMatchSnapshot();
  });
  it('should render correctly changing sortBy', () => {
    useSelector.mockReturnValue({
      filters: [],
      vehicles: mockVehicles,
      sortBy: {
        key: 'doors',
        order: 'asc',
      },
    });
    const { container } = render(<VehicleList />);
    expect(container).toMatchSnapshot();
  });
  it('should render correctly adding filter', () => {
    useSelector.mockReturnValue({
      filters: [{value: 'AVIS', field: 'vendor', label: 'AVIS'}],
      vehicles: mockVehicles,
      sortBy: {
        key: 'doors',
        order: 'asc',
      },
    });
    const { container } = render(<VehicleList />);
    expect(container).toMatchSnapshot();
  });
  it('should dispatch removeFilter tag click', () => {
    useSelector.mockReturnValue({
      filters: [{value: 'AVIS', field: 'vendor', label: 'AVIS'}],
      vehicles: mockVehicles,
      sortBy: {
        key: 'doors',
        order: 'asc',
      },
    });
    const filter = {value: 'AVIS', field: 'vendor'};
    const { getAllByTestId } = render(<VehicleList />);
    const [tag] = getAllByTestId('tag');
    fireEvent.click(tag);
    expect(mockDispatch).toBeCalledWith({
      type: REMOVE_FILTER,
      ...filter,
    });
  });
  it('should dispatch changeSortBy tag click', () => {
    useSelector.mockReturnValue({
      filters: [{value: 'AVIS', field: 'vendor', label: 'AVIS'}],
      vehicles: mockVehicles,
      sortBy: {
        key: 'doors',
        order: 'asc',
      },
    });
    const sort = {key:'doors', order: 'asc'};
    const { getByTestId } = render(<VehicleList />);
    const select = getByTestId('sort-select');
    fireEvent.change(select);
    expect(mockDispatch).toBeCalledWith({
      type: CHANGE_SORT_BY,
      ...sort,
    });
  });
});
