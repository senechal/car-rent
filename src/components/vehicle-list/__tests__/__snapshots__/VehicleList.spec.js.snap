// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing VehicleList should render correctly 1`] = `
<div>
  <section
    class="rac-vehicle-list-header"
  >
    <section
      class="rac-vehicle-list-applied-filters"
    >
      Filters applied: 
    </section>
    <section
      class="rac-vehicle-list-sort-by"
    >
      <label>
         Sort By: 
      </label>
      <select
        data-testid="sort-select"
      >
        <option
          value="price:asc"
        >
          Price: Low to Hight
        </option>
        <option
          value="price:desc"
        >
          Price: Hight to Low
        </option>
        <option
          value="seats:asc"
        >
          Seats
        </option>
        <option
          value="doors:asc"
        >
          Doors
        </option>
        <option
          value="baggageQuantity:asc"
        >
          Bags
        </option>
      </select>
    </section>
  </section>
  <section
    class="rac-vehicle"
  >
    <div
      class="rac-vehicle-content"
    >
      <div
        class="rac-vehicle-img"
      >
        <a
          href="/details/HERTZ_1"
        >
          <img
            alt="Kia Rio or similar"
            src="https://ctimg-fleet.cartrawler.com/kia/rio/primary.png?auto=format&w=600"
          />
        </a>
      </div>
      <div
        class="rac-vehicle-info"
      >
        <div
          class="rac-vehicle-make-model"
        >
          <a
            href="/details/HERTZ_1"
          >
            <strong>
              Kia Rio
            </strong>
             or 
            similar
             (
            Available
            )
          </a>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                seats.svg
              </svg>
            </span>
             
            5
             seats 
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                doors.svg
              </svg>
            </span>
             
            2
             doors 
          </span>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                bag.svg
              </svg>
            </span>
             
            2
             bags 
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                ac.svg
              </svg>
            </span>
             
            Air Conditioning
             
          </span>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                fuel.svg
              </svg>
            </span>
             
            Petrol
             
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                automatic.svg
              </svg>
            </span>
             Automatic Transmission
          </span>
        </div>
      </div>
      <div
        class="rac-vehicle-pricing"
      >
        CA$731.47
      </div>
    </div>
    <div
      class="rac-vehicle-vendor"
    >
      <strong>
        Vendor: 
      </strong>
      <img
        alt="HERTZ"
        src="hertz_logo.gif"
      />
    </div>
  </section>
  <section
    class="rac-vehicle"
  >
    <div
      class="rac-vehicle-content"
    >
      <div
        class="rac-vehicle-img"
      >
        <a
          href="/details/AVIS_0"
        >
          <img
            alt="Chevrolet Impala or similar"
            src="https://ctimg-fleet.cartrawler.com/chevrolet/impala/primary.png?auto=format&w=600"
          />
        </a>
      </div>
      <div
        class="rac-vehicle-info"
      >
        <div
          class="rac-vehicle-make-model"
        >
          <a
            href="/details/AVIS_0"
          >
            <strong>
              Chevrolet Impala
            </strong>
             or 
            similar
             (
            Available
            )
          </a>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                seats.svg
              </svg>
            </span>
             
            5
             seats 
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                doors.svg
              </svg>
            </span>
             
            4
             doors 
          </span>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                bag.svg
              </svg>
            </span>
             
            4
             bags 
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                ac.svg
              </svg>
            </span>
             
            Air Conditioning
             
          </span>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                fuel.svg
              </svg>
            </span>
             
            Petrol
             
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                automatic.svg
              </svg>
            </span>
             Automatic Transmission
          </span>
        </div>
      </div>
      <div
        class="rac-vehicle-pricing"
      >
        CA$836.25
      </div>
    </div>
    <div
      class="rac-vehicle-vendor"
    >
      <strong>
        Vendor: 
      </strong>
      <img
        alt="AVIS"
        src="avis_logo.gif"
      />
    </div>
  </section>
  <section
    class="rac-vehicle"
  >
    <div
      class="rac-vehicle-content"
    >
      <div
        class="rac-vehicle-img"
      >
        <a
          href="/details/ALAMO_1"
        >
          <img
            alt="Chrysler 300 or similar"
            src="https://ctimg-fleet.cartrawler.com/chrysler/300/primary.png?auto=format&w=600"
          />
        </a>
      </div>
      <div
        class="rac-vehicle-info"
      >
        <div
          class="rac-vehicle-make-model"
        >
          <a
            href="/details/ALAMO_1"
          >
            <strong>
              Chrysler 300
            </strong>
             or 
            similar
             (
            Available
            )
          </a>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                seats.svg
              </svg>
            </span>
             
            5
             seats 
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                doors.svg
              </svg>
            </span>
             
            4
             doors 
          </span>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                bag.svg
              </svg>
            </span>
             
            4
             bags 
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                ac.svg
              </svg>
            </span>
             
            Air Conditioning
             
          </span>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                fuel.svg
              </svg>
            </span>
             
            Petrol
             
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                automatic.svg
              </svg>
            </span>
             Automatic Transmission
          </span>
        </div>
      </div>
      <div
        class="rac-vehicle-pricing"
      >
        CA$851.03
      </div>
    </div>
    <div
      class="rac-vehicle-vendor"
    >
      <strong>
        Vendor: 
      </strong>
      <img
        alt="ALAMO"
        src="alamo_logo.gif"
      />
    </div>
  </section>
</div>
`;

exports[`Testing VehicleList should render correctly adding filter 1`] = `
<div>
  <section
    class="rac-vehicle-list-header"
  >
    <section
      class="rac-vehicle-list-applied-filters"
    >
      Filters applied: 
      <span
        class="rac-tag"
        data-tagvalue="vendor:AVIS"
        data-testid="tag"
      >
        AVIS
      </span>
    </section>
    <section
      class="rac-vehicle-list-sort-by"
    >
      <label>
         Sort By: 
      </label>
      <select
        data-testid="sort-select"
      >
        <option
          value="price:asc"
        >
          Price: Low to Hight
        </option>
        <option
          value="price:desc"
        >
          Price: Hight to Low
        </option>
        <option
          value="seats:asc"
        >
          Seats
        </option>
        <option
          value="doors:asc"
        >
          Doors
        </option>
        <option
          value="baggageQuantity:asc"
        >
          Bags
        </option>
      </select>
    </section>
  </section>
  <section
    class="rac-vehicle"
  >
    <div
      class="rac-vehicle-content"
    >
      <div
        class="rac-vehicle-img"
      >
        <a
          href="/details/AVIS_0"
        >
          <img
            alt="Chevrolet Impala or similar"
            src="https://ctimg-fleet.cartrawler.com/chevrolet/impala/primary.png?auto=format&w=600"
          />
        </a>
      </div>
      <div
        class="rac-vehicle-info"
      >
        <div
          class="rac-vehicle-make-model"
        >
          <a
            href="/details/AVIS_0"
          >
            <strong>
              Chevrolet Impala
            </strong>
             or 
            similar
             (
            Available
            )
          </a>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                seats.svg
              </svg>
            </span>
             
            5
             seats 
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                doors.svg
              </svg>
            </span>
             
            4
             doors 
          </span>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                bag.svg
              </svg>
            </span>
             
            4
             bags 
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                ac.svg
              </svg>
            </span>
             
            Air Conditioning
             
          </span>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                fuel.svg
              </svg>
            </span>
             
            Petrol
             
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                automatic.svg
              </svg>
            </span>
             Automatic Transmission
          </span>
        </div>
      </div>
      <div
        class="rac-vehicle-pricing"
      >
        CA$836.25
      </div>
    </div>
    <div
      class="rac-vehicle-vendor"
    >
      <strong>
        Vendor: 
      </strong>
      <img
        alt="AVIS"
        src="avis_logo.gif"
      />
    </div>
  </section>
</div>
`;

exports[`Testing VehicleList should render correctly changing sortBy 1`] = `
<div>
  <section
    class="rac-vehicle-list-header"
  >
    <section
      class="rac-vehicle-list-applied-filters"
    >
      Filters applied: 
    </section>
    <section
      class="rac-vehicle-list-sort-by"
    >
      <label>
         Sort By: 
      </label>
      <select
        data-testid="sort-select"
      >
        <option
          value="price:asc"
        >
          Price: Low to Hight
        </option>
        <option
          value="price:desc"
        >
          Price: Hight to Low
        </option>
        <option
          value="seats:asc"
        >
          Seats
        </option>
        <option
          value="doors:asc"
        >
          Doors
        </option>
        <option
          value="baggageQuantity:asc"
        >
          Bags
        </option>
      </select>
    </section>
  </section>
  <section
    class="rac-vehicle"
  >
    <div
      class="rac-vehicle-content"
    >
      <div
        class="rac-vehicle-img"
      >
        <a
          href="/details/HERTZ_1"
        >
          <img
            alt="Kia Rio or similar"
            src="https://ctimg-fleet.cartrawler.com/kia/rio/primary.png?auto=format&w=600"
          />
        </a>
      </div>
      <div
        class="rac-vehicle-info"
      >
        <div
          class="rac-vehicle-make-model"
        >
          <a
            href="/details/HERTZ_1"
          >
            <strong>
              Kia Rio
            </strong>
             or 
            similar
             (
            Available
            )
          </a>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                seats.svg
              </svg>
            </span>
             
            5
             seats 
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                doors.svg
              </svg>
            </span>
             
            2
             doors 
          </span>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                bag.svg
              </svg>
            </span>
             
            2
             bags 
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                ac.svg
              </svg>
            </span>
             
            Air Conditioning
             
          </span>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                fuel.svg
              </svg>
            </span>
             
            Petrol
             
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                automatic.svg
              </svg>
            </span>
             Automatic Transmission
          </span>
        </div>
      </div>
      <div
        class="rac-vehicle-pricing"
      >
        CA$731.47
      </div>
    </div>
    <div
      class="rac-vehicle-vendor"
    >
      <strong>
        Vendor: 
      </strong>
      <img
        alt="HERTZ"
        src="hertz_logo.gif"
      />
    </div>
  </section>
  <section
    class="rac-vehicle"
  >
    <div
      class="rac-vehicle-content"
    >
      <div
        class="rac-vehicle-img"
      >
        <a
          href="/details/AVIS_0"
        >
          <img
            alt="Chevrolet Impala or similar"
            src="https://ctimg-fleet.cartrawler.com/chevrolet/impala/primary.png?auto=format&w=600"
          />
        </a>
      </div>
      <div
        class="rac-vehicle-info"
      >
        <div
          class="rac-vehicle-make-model"
        >
          <a
            href="/details/AVIS_0"
          >
            <strong>
              Chevrolet Impala
            </strong>
             or 
            similar
             (
            Available
            )
          </a>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                seats.svg
              </svg>
            </span>
             
            5
             seats 
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                doors.svg
              </svg>
            </span>
             
            4
             doors 
          </span>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                bag.svg
              </svg>
            </span>
             
            4
             bags 
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                ac.svg
              </svg>
            </span>
             
            Air Conditioning
             
          </span>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                fuel.svg
              </svg>
            </span>
             
            Petrol
             
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                automatic.svg
              </svg>
            </span>
             Automatic Transmission
          </span>
        </div>
      </div>
      <div
        class="rac-vehicle-pricing"
      >
        CA$836.25
      </div>
    </div>
    <div
      class="rac-vehicle-vendor"
    >
      <strong>
        Vendor: 
      </strong>
      <img
        alt="AVIS"
        src="avis_logo.gif"
      />
    </div>
  </section>
  <section
    class="rac-vehicle"
  >
    <div
      class="rac-vehicle-content"
    >
      <div
        class="rac-vehicle-img"
      >
        <a
          href="/details/ALAMO_1"
        >
          <img
            alt="Chrysler 300 or similar"
            src="https://ctimg-fleet.cartrawler.com/chrysler/300/primary.png?auto=format&w=600"
          />
        </a>
      </div>
      <div
        class="rac-vehicle-info"
      >
        <div
          class="rac-vehicle-make-model"
        >
          <a
            href="/details/ALAMO_1"
          >
            <strong>
              Chrysler 300
            </strong>
             or 
            similar
             (
            Available
            )
          </a>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                seats.svg
              </svg>
            </span>
             
            5
             seats 
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                doors.svg
              </svg>
            </span>
             
            4
             doors 
          </span>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                bag.svg
              </svg>
            </span>
             
            4
             bags 
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                ac.svg
              </svg>
            </span>
             
            Air Conditioning
             
          </span>
        </div>
        <div
          class="rac-vehicle-features"
        >
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                fuel.svg
              </svg>
            </span>
             
            Petrol
             
          </span>
          <span>
            <span
              class="anticon rac-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                focusable="false"
                height="1em"
                width="1em"
              >
                automatic.svg
              </svg>
            </span>
             Automatic Transmission
          </span>
        </div>
      </div>
      <div
        class="rac-vehicle-pricing"
      >
        CA$851.03
      </div>
    </div>
    <div
      class="rac-vehicle-vendor"
    >
      <strong>
        Vendor: 
      </strong>
      <img
        alt="ALAMO"
        src="alamo_logo.gif"
      />
    </div>
  </section>
</div>
`;
